server/parsec/components/memory/datamodel.py

```diff
@@ -227,6 +244,181 @@ def active_sequester_services(self) -> Iterable[MemorySequesterService]:
+
+    def simulate_postgresql_block_table(self) -> Iterable[tuple[int, MemoryBlock]]:
+        """
+        Simulate the PostgreSQL table the blocks are supposed to be stored into.
+
+        This is useful for the realm export feature, since it relies on the table
+        sequential primary key to determine which row should be exported.
+
+        This returns a list of blocks with their sequential primary key.
+        """
+
+        # Here we simulate the sequential primary key of the blocks table in PostgreSQL:
+        # - Blocks are stored in a dict, in Python a dict is guaranteed to be ordered
+        #   according to insertion order.
+        # - We never remove blocks from the dict.
+        # - From the above two points, we can reliably use the index of the block in the
+        #   dict as the primary key.
+
+        # PostgreSQL sequential index starts at 1, however here we skip a bunch of
+        # indexes as a poor man's way to simulate the fact there can be holes in the
+        # indexes (e.g. when a transaction is rolled back).
+        # This should be enough to detect typical improper use of the primary key as
+        # a list offset.
+        return enumerate(self.blocks.values(), start=100)
+
+    def simulate_postgresql_vlob_atom_table(self) -> Iterable[tuple[int, MemoryVlobAtom]]:
+        """
+        Simulate the PostgreSQL table the vlob atoms are supposed to be stored into.
+
+        This is useful for the realm export feature, since it relies on the table
+        sequential primary key to determine which row should be exported.
+
+        This returns a list of vlob atoms with their sequential primary key.
+        """
+
+        # Simulating the sequential primary key of the vlobs table in PostgreSQL is a bit
+        # more tricky than for blocks: we cannot directly rely on the dict since it itself
+        # contains a list of vlob atoms (i.e. a vlob is composed of multiple versions
+        # called atoms).
+        # So we first hove to re-create a list of all vlob atoms across all realms,
+        # sort them by creation date, which is basically equivalent of what the vlobs
+        # table in PostgreSQL is.
+
+        all_vlob_atoms = []
+        for vlob in self.vlobs.values():
+            for vlob_atom in vlob:
+                all_vlob_atoms.append(vlob_atom)
+
+        # Note we also order by vlob ID to ensure a stable order in case of same creation date
+        all_vlob_atoms.sort(key=lambda vlob_atom: (vlob_atom.created_on, vlob_atom.vlob_id))
+
+        # PostgreSQL sequential index starts at 1, however here we skip a bunch of
+        # indexes as a poor man's way to simulate the fact there can be holes in the
+        # indexes (e.g. when a transaction is rolled back).
+        # This should be enough to detect typical improper use of the primary key as
+        # a list offset.
+        return enumerate(all_vlob_atoms, start=200)
```

Comment:
Is there a specific reason to start at 200 here and at 100 in simulate_postgresql_block_table above ?

Reply:
The start value is totally arbitrary since its goal is only to detect typical improper use of the primary key as a list offset.

Having different arbitrary offsets is convenient to help debugging since it makes values more unique.